cmd/lotus-shed/state-stats.go

```diff
@@ -166,49 +186,393 @@ var statObjCmd = &cli.Command{
 			return err
 		}
 
-		r, err := repo.NewFS(cctx.String("repo"))
+		h, err := loadChainStore(ctx, cctx.String("repo"))
 		if err != nil {
-			return xerrors.Errorf("opening fs repo: %w", err)
+			return err
 		}
+		defer h.closer()
 
-		exists, err := r.Exists()
-		if err != nil {
+		dag := merkledag.NewDAGService(blockservice.New(h.bs, offline.Exchange(h.bs)))
+		dsc := &dagStatCollector{
+			ds:   dag,
+			walk: carWalkFunc,
+		}
+
+		if err := merkledag.Walk(ctx, dsc.walkLinks, c, cid.NewSet().Visit, merkledag.Concurrent()); err != nil {
 			return err
 		}
-		if !exists {
-			return xerrors.Errorf("lotus repo doesn't exist")
+
+		return DumpJSON(dsc.stats)
+	},
+}
+
+type StoreHandle struct {
+	bs     blockstore.Blockstore
+	cs     *store.ChainStore
+	sm     *stmgr.StateManager
+	closer func()
+}
+
+func loadChainStore(ctx context.Context, repoPath string) (*StoreHandle, error) {
```

Comment:
Probably could have a mode where this loads a snapshot to a memory-map blockstore, would need a decent amount of RAM, but should be much faster